namcap.1
.TH namcap 1 "Jan 13, 2024" "namcap 3.5.1" "User Commands"
.SH NAME
namcap \- package analysis utility
.SH SYNOPSIS
\fBnamcap [options] <package|PKGBUILD> [package|PKGBUILD] ...
.SH DESCRIPTION
.PP
\fBnamcap\fP is a \fIpackage analysis\fP utility that looks for problems with Arch Linux packages or their PKGBUILD files.  It can apply rules to the file list, the files themselves, or individual PKGBUILD files.
.PP
Rules return lists of messages.  Each message can be one of three types: error, warning, or information (think of them as notes or comments).  Errors (designated by 'E:') are things that namcap is very sure are wrong and need to be fixed.  Warnings (designated by 'W:') are things that namcap thinks should be changed but if you know what you're doing then you can leave them.  Information (designated 'I:') are only shown when you use the info argument.  Information messages give information that might be helpful but isn't anything that needs changing.
.SH OPTIONS
.TP
\fB\-e\fR RULELIST, \fB\-\-exclude=\fRRULELIST
Do not run RULELIST rules on the package
.TP
.B "\-i, \-\-info"
display information messages
.TP
.B "\-L, \-\-list
return a list of valid rules and their descriptions
.TP
.B "\-m, \-\-machine\-readable"
displays easily parseable namcap tags instead of the normal human readable description; for example using non-fhs-man-page instead of "Non-FHS man page (%s) found. Use /usr/share/man instead". A full list of namcap tags along with their human readable descriptions can be found at /usr/share/namcap/tags.
.TP
\fB\-r\fR RULELIST, \fB\-\-rules=\fRRULELIST
only apply RULELIST rules to the package
.IP
RULELIST is a comma-separated list of rule names
.TP
.B "\-v, \-\-version"
print version and exit
.SH RULES
.TP
.B arrays
Verifies that array variables are actually arrays and not strings
.TP
.B badbackups
Verifies that backup entries don't begin with /
.TP
.B carch
Looks for references to i686 and i586 and tells you to use $CARCH instead
.TP
.B capsnames
Checks a PKGBUILD to verify that the package name has no upper case characters
.TP
.B capsnamespkg
Checks a package to verify that the package name has no upper case characters
.TP
.B checksums
Makes sure that a PKGBUILD includes valid checksums
.TP
.B depends
This module runs ldd on all executables, gets the link-level dependencies, finds the smallest subset of dependencies that cover the link-level dependencies, and compares that list to the depends of the package.  It returns messages in three cases: dependency detected and not included, dependency included but already satisfied, and dependency included and not needed.  These suggestions are just guidelines and all package builders should take this into account (i.e. you're smarter than namcap is)

Some cases where namcap fails are dlopen() and obscure links.  dlopen()'d libraries don't show up because they are loaded at run time: in the case of a program that loads plugins.  Obscure links are the cases where only a small portion of the package needs something to run; usually, the small portion won't be accessed unless that thing is installed (i.e. a java plugin for mozilla included in the java package).
.TP
.B directoryname
Checks the directory names in the package file list for standard directories (i.e. /etc, /usr/bin, /usr/lib).  All files in directories that are non-standard are returned
.TP
.B elffiles
Gives an error if it finds ELF (binary) files outside the standard paths (/bin, /usr/bin, etc.)
.TP
.B emptydir
Looks for directories in a package that don't contain any files
.TP
.B extravars
Verifies that extra variables start with an underscore
.TP
.B fhsinfopages
Verifies correct installation of info pages
.TP
.B fhsmanpages
Verifies correct installation of man pages
.TP
.B fileownership
Verifies that all files are owned by root:root and not some other user or group
.TP
.B gnomemime
Checks for generated GNOME mime files
.TP
.B hicoloricons
Checks whether the hicolor icon cache is updated when a package installs
files in /usr/share/icons/hicolor
.TP
.B infodirectory
Verifies that a package does not contain the info directory file (/usr/share/info/dir)
.TP
.B invalidstartdir
Any PKGBUILDs that refer to files in $startdir are incorrect
.TP
.B libtool
Warns if libtool (*.la) files are present
.TP
.B license
Verifies that the licenses variable has been filled in in a PKGBUILD.
.TP
.B licensepkg
Verifies that the licenses variable has been filled in in a package. For packages with a custom license it
checks whether the license file has been installed in
/usr/share/licenses/$pkgname/
.TP
.B perllocal
Searches for perllocal.pod.  perllocal.pod is a nasty file that's included during most perl module installations
.TP
.B permissions
Checks basic file and and directory permissions.  It returns warnings about world readable bits not being set, world writable bits being set, and world executable bit not being set for directories
.TP
.B rpath
Gives an error if a binary has RPATH set to something other than /usr/lib
.TP
.B runpath
Gives an error if a binary has RUNPATH set to something other than /usr/lib, /usr/lib32
.TP
.B scrollkeeper
Verifies that there aren't any scrollkeeper directories
.TP
.B sfurl
Warns if a PKGBUILD is downloading from a specific sourceforge mirror instead of the generic downloads.sourceforge.net
.TP
.B symlink
Finds out symbolic and hard links which do not point to a valid path in the package
.TP
.B tags
For package maintainers, this module verifies that we have remembered to add a Maintainer and Contributor comment.
.TP
.B url
Verifies that we have the url variable set in the PKGBUILD
.TP
.B urlpkg
Verifies that we have the url variable set in the package file
.SH EXAMPLES
.TP
.B namcap foo.pkg.tar.gz
apply all rules to foo.pkg.tar.gz
.TP
.B namcap -r depends,symlink foo.pkg.tar.gz
apply the depends and symlink rules to foo.pkg.tar.gz
.TP
.B namcap PKGBUILD
apply all PKGBUILD based rules to the file PKGBUILD
.TP
.B namcap --list
list all of the available rules
.SH COPYRIGHT
Copyright \(co 2003-2023 Namcap contributors, see AUTHORS for details.
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.fi


NEWS
namcap (3.5.1)
    - Fix message handling from license detection
    - Fix unpacking of packages to check symlinked licenses
    - Add missing info to license file missing message
    - Add type annotations detect issues when adding messages
    - Fix crash while parsing invalid shebang lines

namcap (3.5.0)
    - Validate license field using SPDX identifiers
    - Check for shadow stack support
    - And QML dependency checks

namcap (3.4.2)
    - Revert incorrectly dropped rule imports

namcap (3.4.1)
    - Use .pc files from packages not matching ones from host system
    - Use raw strings to disambiguate regular expressions
    - Copy-edit wording of tags
    - Fix import path scopeing in tests
    - Refresh licensing, author lists, testing tips, and other info in docs
    - Reformat and lint entire codebase (via black & ruff)

namcap (3.4.0)
    - Overhaul Python dependency handling
    - Overhaul sodepends including architecture pinning and RPATH handling
    - Add warning for transitive dependencies
    - Add color to make errors and warnings easier to parse if output is TTY
    - Add check for dependencies called out in pkgconfig files
    - Fix handling of debug packages
    - Improve error and warning message content
    - Improve licence handling especially of linked licenses and empty arrays
    - Drop obsolete KDE specific rules
    - Refactor Python coding to remove deprecated features, improve readability
    - Implement CI testing and coverage reporting

namcap (3.3.1)
    - Avoid failure when scanning dependencies in invalid Python files

namcap (3.3.0)
    - Detect Python dependencies, Python extensions with limited ABIs
    - Support Python 3.10
    - Detect binaries that may execute code from shared RUNPATH
    - Replace bzr with breezy
    - Fix OFL license handling
    - Handle debug packages (by ignoring some checks)
    - Handle source array being empty
    - Accommodate project development on case-insensitive filesystems

namcap (3.2.10)
    - make tests pass by fixing the version in the manpage

namcap (3.2.9)
    - warn for duplicated source filenames
    - improved split package handling
    - warn if depends duplicated in makedeps
    - improved VCS handling
    - improved architecture specific var handling
    - remove warnings handled by hooks (FS#56898)
    - add warning for unneeded hook deps
    - new compression methods in pacman 5.2 (FS#64570)

namcap (3.2.8)
    - fix java-runtime handling (FS#24296)
    - stricter shebang parsing (FS#49733)
    - /usr/src is okay (FS#49889)
    - fix API changes in pyelftools 0.24 (FS#50428)
    - summarize ELF files found in opt (FS#49489)
    - remove use of eu-findtextrel (FS#42700)
    - check for RELRO in ELF files (FS#26435)
    - warn on use of undocumented makepkg functions (FS#43502)
    - more robust parsing (FS#53161)
    - packages can symlink to files in depenencies (FS#53282)
    - warn on unstripped files (FS#27485)
    - better file handling (FS#57193)
    - FULL RELRO check
    - much cleaning and refactoring
    - many updates to variables/hooks

namcap (3.2.7)
    - improved carch depends (FS#47862)
    - better multilib dependencies
    - detects bad py/pyc/pyo mtimes
    - detects when an install file shouldn't be used
    - removed old install file rules

namcap (3.2.6)
    - better mingw support (FS#43552)
    - lrz/lzo compression (FS#44900)
    - handle validpgpkeys warnings
    - improved carch-specific depends (FS#45455, FS#45456)
    - test for overlinking

namcap (3.2.5)
    - support "SKIP" in checksums (FS#34647)
    - export CARCH correctly to parsepkgbuild (FS#32568)
    - recognize .MTREE as package metadata (FS#34591)

namcap (3.2.4)
    - update fhs rules for new guidelines (/usr move).

namcap (3.2.3)
    - add pyelftools as a dependency, drop prelink.
    - add rubypaths rule warning against use of site_ruby
    - modify elfexecstack rule to use elftools
    - add test case for elfexecstack

namcap (3.2.2)
    - fix: unwanted warning when testing database is not found (FS#27973)
    - fix: Creative Commons licenses not recognized (FS#27489)
    - rename elffiles rule to elfpaths
    - add elftextrel rule to detect text relocations (FS#26434)
    - add elfexecstack rule to detect executable stacks (FS#26458)

namcap (3.2.1)
    - fix: glibfiles rules ineffective because of wrong schema
      filename pattern
    - fix: incorrect warning about permissions for symbolic links to
      static libraries (FS#26536)

namcap (3.2)
    - requires pacman 4 / pyalpm 0.5
    - fix: consider archive files as binaries in anyelf (FS#24854)
    - fix: refine glibfiles rules to avoid warnings in glib2 (FS#25829)

namcap (3.1.1)
    - fix: carch rule behavior
    - various code cleanups

namcap (3.1)
    - add checks for files in temp. folders in fhs rule (FS#23539)
    - add rule javafiles detecting Java class/jar files (FS#23521)
    - fix: parsepkgbuild executing arbitrary commands from PKGBUILD (FS#23902)
    - fix: extravars not knowing about $epoch (FS#23966)
    - fix: mimedesktop rule crashing on packages with symlinks (FS#23952)
    - fix: bug in script interpreter detection
    - uses pyalpm to access pacman database

namcap (3.0.6)
    - depends check now prints reasons why a dependency is needed
    - fix: inverted treatment of provides introduced by 3.0.5
    - test: basic unit test for sodepends rule
    - factored out tag formatting code

namcap (3.0.5)
    - fix: depends rule unreadable and broken (FS#23489)
    - added a warning when a rule needs a dependency declared optional

namcap (3.0.4)
    - run parsepkgbuild with a clean environment
    - added checkdepends to the list of legit PKGBUILD variables

namcap (3.0.3)
    - fix: extravars rule having unstable behaviour
    - fix: symlink rule not handling particular symlinks

namcap (3.0.2)
    - extravars rule: reduce the amount of false positives
    - fix crashes when running in non-English locales

namcap (3.0.1)

namcap.py
#!/usr/bin/env python3
# Copyright (C) 2003-2023 Namcap contributors, see AUTHORS for details.
# SPDX-License-Identifier: GPL-2.0-or-later

import getopt
import os
import sys
import tarfile

import Namcap.depends
import Namcap.rules
import Namcap.tags
import Namcap.version


# Functions
def get_modules():
    """Return all possible modules (rules)"""
    return Namcap.rules.all_rules


def usage():
    """Display usage information"""
    print("")
    print("Usage: " + sys.argv[0] + " [OPTIONS] packages")
    print("")
    print("Options are:")
    print("    -L, --list                       : list available rules")
    print("    -i                               : prints information (debug) responses from rules")
    print("    -m                               : makes the output parseable (machine-readable)")
    print("    -e rulelist, --exclude=rulelist  : don't apply RULELIST rules to the package")
    print("    -r rulelist, --rules=rulelist    : only apply RULELIST rules to the package")
    print("    -t tags                          : use a custom tag file")
    print("    -v version                       : print version and exit")

    sys.exit(2)


def open_package(filename):
    try:
        tar = tarfile.open(filename, "r")
        if ".PKGINFO" not in tar.getnames():
            tar.close()
            return None
    except IOError:
        if tar:
            tar.close()
        return None
    return tar


def check_rules_exclude(optlist):
    """Check if the -r (--rules) and the -r (--exclude) options
    are being used at same time"""
    args_used = 0
    for i in optlist:
        if "-r" in i or "-e" in i:
            args_used += 1
        if "--rules" in i or "--exclude" in i:
            args_used += 1
    return args_used


def show_messages(name, key, messages):
    colored_key = {
        "E": "\033[91mE\033[00m",
        "W": "\033[93mW\033[00m",
        "I": "\033[92mI\033[00m",
    }
    for msg in messages:
        if sys.stdout.isatty():
            print("%s %s: %s" % (name, colored_key[key], Namcap.tags.format_message(msg)))
        else:
            print("%s %s: %s" % (name, key, Namcap.tags.format_message(msg)))


def process_realpackage(package, modules):
    """Runs namcap checks over a package tarball"""
    pkgtar = open_package(package)

    if not pkgtar:
        print("Error: %s is empty or is not a valid package" % package)
        return 1

    pkginfo = Namcap.package.load_from_tarball(package)
    # Loop through each one, load them apply if possible
    for i in modules:
        rule = get_modules()[i]()

        if isinstance(rule, Namcap.ruleclass.PkgInfoRule):
            rule.analyze(pkginfo, None)
        elif isinstance(rule, Namcap.ruleclass.PkgbuildRule):
            pass
        elif isinstance(rule, Namcap.ruleclass.TarballRule):
            rule.analyze(pkginfo, pkgtar)
        else:
            show_messages(pkginfo["name"], "E", [("error-running-rule %s", i)])

        # Output the three types of messages
        show_messages(pkginfo["name"], "E", rule.errors)
        show_messages(pkginfo["name"], "W", rule.warnings)
        if info_reporting:
            show_messages(pkginfo["name"], "I", rule.infos)

    # dependency analysis
    errs, warns, infos = Namcap.depends.analyze_depends(pkginfo)
    show_messages(pkginfo["name"], "E", errs)
    show_messages(pkginfo["name"], "W", warns)
    if info_reporting:
        show_messages(pkginfo["name"], "I", infos)


def process_pkginfo(pkginfo, modules):
    """Runs namcap checks of a single, non-split PacmanPackage object"""
    for i in modules:
        rule = get_modules()[i]()
        if isinstance(rule, Namcap.ruleclass.PkgInfoRule):
            rule.analyze(pkginfo, None)

        # Output the messages
        if "base" in pkginfo:
            name = "PKGBUILD (" + pkginfo["base"] + ")"
        else:
            name = "PKGBUILD (" + pkginfo["name"] + ")"
        show_messages(name, "E", rule.errors)
        show_messages(name, "W", rule.warnings)
        if info_reporting:
            show_messages(name, "I", rule.infos)


def process_pkgbuild(package, modules):
    """Runs namcap checks over a PKGBUILD"""
    # We might want to do some verifying in here... but really... isn't that
    # what pacman.load is for?
    pkginfo = Namcap.package.load_from_pkgbuild(package)

    if pkginfo is None:
        print("Error: %s is not a valid PKGBUILD" % package)
        return 1

    # apply global PKGBUILD rules
    for i in modules:
        rule = get_modules()[i]()
        if isinstance(rule, Namcap.ruleclass.PkgbuildRule):
            rule.analyze(pkginfo, package)
        # Output the messages
        if "base" in pkginfo:
            name = "PKGBUILD (" + pkginfo["base"] + ")"
        else:
            name = "PKGBUILD (" + pkginfo["name"] + ")"
        show_messages(name, "E", rule.errors)
        show_messages(name, "W", rule.warnings)
        if info_reporting:
            show_messages(name, "I", rule.infos)
    # apply per pkginfo rule
    for subpkg in pkginfo.subpackages if pkginfo.is_split else [pkginfo]:
        process_pkginfo(subpkg, modules)


# Main
modules = get_modules()
info_reporting = 0
machine_readable = False
filename = None

# get our options and process them
try:
    optlist, args = getopt.getopt(
        sys.argv[1:],
        "ihmr:e:t:Lv",
        [
            "info",
            "help",
            "machine-readable",
            "rules=",
            "exclude=",
            "tags=",
            "list",
            "version",
        ],
    )
except getopt.GetoptError:
    usage()

active_modules = {}

# Verifying if we are using the -r and -r options at same time
if check_rules_exclude(optlist) > 1:
    print("You cannot use '-r' (--rules) and '-e' (-exclude) options at same time")
    usage()

for i, k in optlist:
    if i in ("-L", "--list"):
        print("-" * 20 + " Namcap rule list " + "-" * 20)
        for j in sorted(modules):
            print("%-20s: %s" % (j, modules[j].description))
        sys.exit(2)

    if i in ("-r", "--rules"):
        module_list = k.split(",")
        for j in module_list:
            if j in modules:
                active_modules[j] = modules[j]
            else:
                print("Error: Rule '%s' does not exist" % j)
                usage()

    # Used to exclude some rules from the check
    if i in ("-e", "--exclude"):
        module_list = k.split(",")
        active_modules.update(modules)
        for j in module_list:
            if j in modules:
                active_modules.pop(j)
            else:
                print("Error: Rule '%s' does not exist" % j)
                usage()

    if i in ("-i", "--info"):
        info_reporting = 1

    if i in ("-h", "--help"):
        usage()
    if i in ("-m", "--machine-readable"):
        machine_readable = True

    if i in ("-t", "--tags"):
        filename = k

    if i in ("-v", "--version"):
        print(Namcap.version.get_version())
        sys.exit(0)

# If there are no args, print usage
if args == []:
    usage()

Namcap.tags.load_tags(filename=filename, machine=machine_readable)

packages = args

# No rules selected?  Then select them all!
if len(active_modules) == 0:
    active_modules = modules

# Go through each package, get the info, and apply the rules
for package in packages:
    if not os.access(package, os.R_OK):
        print("Error: Problem reading %s" % package)
        usage()

    if os.path.isfile(package) and tarfile.is_tarfile(package):
        process_realpackage(package, active_modules)
    elif "PKGBUILD" in package:
        process_pkgbuild(package, active_modules)
    else:
        print("Error: %s not package or PKGBUILD" % package)
